numpyro/infer/elbo.py

```diff
@@ -526,24 +529,59 @@ def _compute_downstream_costs(model_trace, guide_trace, non_reparam_nodes):
     return downstream_costs, downstream_guide_cost_nodes
 
 
+class track_nonreparam(Messenger):
+    def postprocess_message(self, msg):
+        # track non-reparameterizable sample sites
+        if (
+            msg["type"] == "sample"
+            and (not msg["is_observed"])
+            and (not msg["fn"].has_rsample)
+        ):
+            new_provenance = frozenset({msg["name"]})
+            old_provenance = msg["value"].aval.named_shape.get(
+                "_provenance", frozenset()
+            )
+            msg["value"].aval.named_shape["_provenance"] = (
+                old_provenance | new_provenance
+            )
+
+
+def get_importance_log_probs(model, guide, args, kwargs, params):
+    """
+    Returns log probabilities at each site for the guide and the model that is run against it.
+    """
+    model_tr, guide_tr = get_importance_trace(model, guide, args, kwargs, params)
```

Comment:
Changed since yesterday: Instead of manually tracing guide and then replaying/tracing model I changed this to just reuse the `get_importance_trace` function.